File: lib/geometry/rectangle.rb
require_relative 'ffi/rectangle'

module Geometry
  class Rectangle
    def initialize(width, height)
      @ptr = FFI::Rectangle.rectangle_make(width, height)
    end

    def area
      FFI::Rectangle.rectangle_area(@ptr)
    end

    def width
      FFI::Rectangle.rectangle_get_width(@ptr)
    end

    def width=(width_to_set)
      FFI::Rectangle.rectangle_set_width(@ptr, width_to_set)
    end

    def height
      FFI::Rectangle.rectangle_get_height(@ptr)
    end

    def height=(height_to_set)
      FFI::Rectangle.rectangle_set_height(@ptr, height_to_set)
    end
  end
end

File: src/rectangle.rs
pub struct Rectangle {
    pub width: f32,
    pub height: f32
}

impl Rectangle {
    pub fn area(&self) -> f32 {
        self.width * self.height
    }
}

// Public API for Rectangle operations

// Initialize
#[no_mangle]
pub extern fn rectangle_make(width: f32, height: f32) -> Box<Rectangle> {
    Box::new(Rectangle { width: width, height: height })
}

// Getters and setters
#[no_mangle]
pub extern fn rectangle_get_width(rectangle: &Rectangle) -> f32 {
    rectangle.width
}

#[no_mangle]
pub extern fn rectangle_set_width(rectangle: &mut Rectangle, width: f32) {
    rectangle.width = width
}

#[no_mangle]
pub extern fn rectangle_get_height(rectangle: &Rectangle) -> f32 {
    rectangle.height
}

#[no_mangle]
pub extern fn rectangle_set_height(rectangle: &mut Rectangle, height: f32) {
    rectangle.height = height
}

// Methods
#[no_mangle]
pub extern fn rectangle_area(rectangle: &Rectangle) -> f32 {
    rectangle.area()
}
